4.analyze_data/notebooks/UMAP/nbconverted/2.find_sc.py

```diff
@@ -0,0 +1,217 @@
+    filtered_df = filtered_df[selected_columns]
+
+    # Randomly select a row
+    random_row = filtered_df.sample(n=1, random_state=info["Random state"])  
+
+    # Add the selected row to the DataFrame
+    df = df.append(random_row, ignore_index=True)
+
+    # Assign the direction to the UMAP_direction column for the last added row
+    df.at[df.index[-1], "UMAP_direction"] = crop_cell.split('-')[2]
+
+    # Save the updated DataFrame to the CSV file
+    df.to_csv(csv_file_path, index=False)
+
+    # Create a filename based on Metadata_Plate, Metadata_Well, Metadata_Site
+    plate = random_row['Metadata_Plate'].values[0]
+    well = random_row['Metadata_Well'].values[0]
+    site = random_row['Metadata_Site'].values[0]
+
+    # Initialize a list to store file paths
+    file_paths = []
+
+    # Create 5 different file paths with "d0" through "d4" suffixes
+    for i in range(5):
+        filename = f"{images_dir}/{plate}_{well}{site}d{i}_illumcorrect.tiff"
+        file_paths.append(filename)
+
+    print("Randomly selected row:")
+    print(random_row)
+    print("Generated filenames:")
+    for path in file_paths:
+        print(path)
+
+    # Initialize empty lists to store the images for each channel
+    blue_channel = []
+    green_channel = []
+    red_channel = []
+
+    # Iterate over files in the directory
+    for file_path in file_paths:
+        filename = pathlib.Path(file_path).name
+        if 'd0' in filename:
+            blue_channel_image = cv2.imread(str(file_path), cv2.IMREAD_UNCHANGED)
+            blue_channel.append(blue_channel_image)
+        elif 'd1' in filename:
+            green_channel_image = cv2.imread(str(file_path), cv2.IMREAD_UNCHANGED)
+            green_channel.append(green_channel_image)
+        elif 'd4' in filename:
+            red_channel_image = cv2.imread(str(file_path), cv2.IMREAD_UNCHANGED)
+            red_channel.append(red_channel_image)
+
+    # Stack the images for each channel along the channel axis
+    blue_channel_stack = np.stack(blue_channel, axis=-1)
+    green_channel_stack = np.stack(green_channel, axis=-1)
+    red_channel_stack = np.stack(red_channel, axis=-1)
+
+    # Scale the pixel values to fit within the 16-bit range (0-65535)
+    blue_channel_stack = (blue_channel_stack / np.max(blue_channel_stack) * 65535).astype(np.uint16)
+    green_channel_stack = (green_channel_stack / np.max(green_channel_stack) * 65535).astype(np.uint16)
+    red_channel_stack = (red_channel_stack / np.max(red_channel_stack) * 65535).astype(np.uint16)
+
+    # Create the RGB image by merging the channels
+    composite_image = cv2.merge((red_channel_stack, green_channel_stack, blue_channel_stack)).astype(np.uint16)
+
+    # Path for saving comp images
+    comp_path = pathlib.Path(f"{comp_dir}/{info['composite_save_path']}")
+
+    # Save the composite 16-bit RGB tiff image
+    tf.imwrite(comp_path, composite_image)
+
+    # Load the composite image
+    composite_image = Image.open(comp_path)
```

Comment:
Do you need to reload the composite image if we just saved it from an object a few. lines above? Does loading the image through a different function change the object type? Maybe add some comments explaining this?

Reply:
Good catch! 

I just looked into it more and the reason why this happens is because we are using pillow to perform the cropping and the image must be of an `Image` object. The current composite_image object prior to loading it back in is a numpy array and I haven't found a good way to get this to work. I have edited the comments to make sure this more clear.